src/mlpack/methods/reinforcement_learning/environment/ftn.hpp

```diff
@@ -0,0 +1,300 @@
+/**
+ * @file methods/reinforcement_learning/environment/ftn.hpp
+ * @author Nanubala Gnana Sai
+ *
+ * This file is an implementation of Fruit Tree Navigation (FTN) Task:
+ * https://github.com/RunzheYang/MORL/blob/master/synthetic/envs/fruit_tree.py
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_ENVIRONMENT_FTN_HPP
+#define MLPACK_METHODS_RL_ENVIRONMENT_FTN_HPP
+
+#include <mlpack/core.hpp>
+
+namespace mlpack{
+namespace rl{
+
+/**
+ * Implementation of Fruit Tree Navigation Task. 
+ */
+class FTN
+{
+ public:
+  /*
+   * Implementation of Fruit Tree Navigation state. Each State is a tuple 
+   * {(row, column)} representing zero based index of a node in the tree.
+   */
+  class State
+  {
+   public:
+    /**
+     * Construct a state instance.
+     */
+    State(): data(dimension) { /* nothing to do here */ }
+
+    /**
+     * Construct a state instance from given data.
+     *
+     * @param data Data for the zero based index of the current node.
+     */
+    State(const arma::colvec& data) : data(data)
+    { /* nothing to do here */ }
+
+    //! Modify the state representation.
+    arma::vec& Data() { return data; }
+
+    //! Get value of row index of the node.
+    double Row() const { return data[0]; }
+    //! Modify value of row index of the node.
+    double& Row() { return data[0]; }
+
+    //! Get value of column index of the node.
+    double Column() const { return data[1]; }
+    //! Modify value of column index of the node.
+    double& Column() { return data[1]; }
+
+    //! Dimension of the encoded state.
+    static constexpr size_t dimension = 2;
+
+   private:
+    //! Locally-Stored {(row, column)}.
+    arma::vec data;
+  };
+
+  /*
+   * Implementation of action for Fruit Tree Navigation task.
+   */
+  class Action
+  {
+   public:
+    enum actions
+    {
+      left,
+      right,
+    };
+    // To store the action.
+    Action::actions action;
+
+    // Track the size of the action space.
+    static const size_t size = 2;
+  };
+
+  /**
+   * Construct a Fruit Tree Navigation instance using the given constants.
+   *
+   * @param maxSteps The number of steps after which the episode
+   *    terminates. If the value is 0, there is no limit.
+   * @param depth The maximum depth of the full binary tree.
+   */
+  FTN(const size_t maxSteps = 500,
```

Comment:
Another minor comment from my side, it could be better if we used the complete name for the class.
```suggestion
  FruitTreeNavigation(const size_t maxSteps = 500,
```